db.go

```diff
@@ -0,0 +1,231 @@
+type DBLike interface {
+	NewInsert() *bun.InsertQuery
+	NewSelect() *bun.SelectQuery
+	NewUpdate() *bun.UpdateQuery
+}
+
+// DB is our OO-style database interface
+// Both RootDB and txDB implement this, but:
+//  - RootDB has a .RunInTx() method that results in a txDB
+//  - txDB is not exposed
+type DB interface {
+	UpsertBuild(b Build) error
+	GetBuild(buildNum int) (*Build, error)
+	GetBuildsForPull(pull int) ([]Build, error)
+	OldestUnfinishedBuild() (*Build, error)
+	LatestFinishedBefore(before *Build) (*Build, error)
+	GetArchivableBuild() (*Build, error)
+	UnarchivedBuilds() ([]Build, error)
+	ArchiveBuild(buildNum int) error
+	UpsertArtifacts(artifacts []Artifact) error
+	GetArtifacts(buildNum int) ([]Artifact, error)
+}
+
+
+// dbImpl implements our OO-style database interface
+// dbImpl automatically implements DBLike by composition.
+type dbImpl struct {
+	DBLike
+}
+
+// RootDB implements DBLike and dbImpl.
+// Additionally RootDB has a RunInTx() method.
+type RootDB struct {
+	// bun.DB makes RootDB implement DBLike
+	*bun.DB
+	// dbImpl will just contain another pointer to the *bun.DB
+	dbImpl
+}
+
+// RootDB.RunInTx() wraps bun.DB.RunInTx to provide the DB interface without giving access to the
+// actual bun.Tx object.
+func (db *RootDB) RunInTx(fn func(DB) error) error {
+	bunFn := func(_ context.Context, tx bun.Tx) error {
+		txdb := dbImpl{&tx}
+		return fn(txdb)
+	}
+	return db.DB.RunInTx(context.Background(), nil, bunFn)
+}
+
+func NewDB(dbPath string) (RootDB, error) {
+	sqldb, err := sql.Open("sqlite3", dbPath)
+	if err != nil {
+		return RootDB{}, err
+	}
+
+	// create the database if it doesn't exist
+	_, err = sqldb.Exec(`
+		CREATE TABLE IF NOT EXISTS builds (
+			build_num INTEGER PRIMARY KEY,
+			url TEXT NOT NULL,
+			branch TEXT NOT NULL,
+			subject TEXT NOT NULL,
+			'commit' TEXT NOT NULL,
+			parallel INTEGER NOT NULL,
+			workflow TEXT,
+			start_time TEXT NOT NULL,
+			lifecycle TEXT NOT NULL,
+			-- this is our own metadata
+			archived BOOLEAN NOT NULL DEFAULT FALSE
+		);
+		CREATE TABLE IF NOT EXISTS artifacts (
+			url TEXT PRIMARY KEY,
+			build_num INTEGER NOT NULL
+		);
+	`)
+	if err != nil {
+		sqldb.Close()
+		return RootDB{}, errors.Wrap(err, "creating tables")
+	}
+
+	db := bun.NewDB(sqldb, sqlitedialect.New())
+	// db.AddQueryHook(bundebug.NewQueryHook(bundebug.WithVerbose(true)))
+	db.AddQueryHook(bundebug.NewQueryHook())
+	return RootDB{db, dbImpl{db}}, nil
+}
+
+func (db dbImpl) UpsertBuild(b Build) error {
+	q := db.NewInsert().
+		Model(&b).
+		ExcludeColumn("archived").
+		On("CONFLICT (build_num) DO UPDATE").
+		Set("lifecycle = EXCLUDED.lifecycle")
+	_, err := q.Exec(context.Background())
+	return err
+}
+
+func (db dbImpl) GetBuild(buildNum int) (*Build, error) {
+	b := Build{BuildNum: buildNum}
+	q := db.NewSelect().
+		Model(&b).
+		WherePK()
+	err := q.Scan(context.Background())
+	if err == sql.ErrNoRows {
+		return nil, nil
+	}
+	if err != nil {
+		return nil, err
+	}
```

Comment:
chain everything / switch, + elsewhere
```suggestion
	err := db.NewSelect().
		Model(&b).
		WherePK().
		Scan(context.Background())
	switch {
	case err == sql.ErrNoRows:
		return nil, nil
	case err != nil:
		return nil, err
	}
```

Reply:
what's the point of the switch?

Reply:
style, all error handling in a single block.

Reply:
didn't take this exact advice, but I did simplify it so to meet the requirement of "one block of error handling"